File: regi-headless/src/main/java/usace/rowcps/headless/RegiCLI.java
package usace.rowcps.headless;

import hec.db.DbConnectionException;
import hec.db.DbIoException;
import hec.db.DbPluginNotFoundException;
import hec.db.InvalidDbConnectionException;
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.context.Scope;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.kohsuke.args4j.CmdLineException;
import usace.rowcps.headless.interfaces.ScriptEvaluator;
import usace.rowcps.regi.factories.RowcpsExecutorService;
import usace.rowcps.regi.model.ManagerId;
import usace.rowcps.regi.model.RegiDomain;

public class RegiCLI
{

	private static final Logger LOGGER = Logger.getLogger(RegiCLI.class.getName());

	public static void main(String[] args)
	{
		Span rootSpan = GlobalOpenTelemetry.getTracer("regi-headless")
			.spanBuilder("runHeadless")
			.startSpan();
		try(Scope scope = rootSpan.makeCurrent())
		{
			runHeadlessTest(args);
		}
		catch (DbConnectionException | DbPluginNotFoundException | IOException | RuntimeException ex)
		{
			rootSpan.recordException(ex);
			rootSpan.setStatus(StatusCode.ERROR);
			LOGGER.log(Level.SEVERE, "Headless error connecting to database.", ex);
			System.exit(-1);
			return;
		}

		LOGGER.info("Exiting.");
		System.exit(0);
	}
	
	/**
	 * Used by TestHeadless unit test class to run headless without calling System.exit(0)
	 * 
	 * @param unused
	 * @throws DbConnectionException
	 * @throws InvalidDbConnectionException
	 * @throws CmdLineException
	 * @throws DbPluginNotFoundException 
	 */
	static void runHeadlessTest(String[] unused)
        throws DbConnectionException, DbPluginNotFoundException,
        IOException {
		
		HeadlessRegiDomainFactory factory = new HeadlessRegiDomainFactory();
		RegiDomain regiDomain = factory.createDomain();
		
		if (regiDomain != null)
		{
			ScriptEvaluator pe = new PythonEvaluator();
			Map<String, Object> vars = new HashMap<>();

			ManagerId managerId = factory.getManagerId();
			RegiCalcRegistry reg = new RegiCalcRegistry(regiDomain, managerId);
			vars.put("registry", reg);
			
			File scriptFile = new File(System.getenv("SCRIPT"));
			
			try
			{
				FileReader fr = new FileReader(scriptFile);
				LOGGER.info("Evaluating script file");
				Object retval = pe.evaluateExpression(fr, vars);
				
				LOGGER.info("Jython script completed normally, commiting data.");
				regiDomain.commitData(managerId);
				LOGGER.info("RegiDomain committed data.");
			}
			catch (DbConnectionException | DbIoException | FileNotFoundException ex)
			{
				LOGGER.log(Level.SEVERE, "Exception occurred while evaluating Jython file:" + System.lineSeparator() + scriptFile, ex);
			}
			finally
			{
				LOGGER.info("RegiDomain closing.");
				shutdownRowcpsAccessFactory(managerId);
				regiDomain.closing();
			}
		}
	}

	private static void shutdownRowcpsAccessFactory(ManagerId managerId)
	{
		LOGGER.log(Level.INFO, "Shutting down RowcpsExecutorService for {0}", managerId);
		RowcpsExecutorService res = RowcpsExecutorService.getInstance(managerId);

		res.shutdown();  // signal shutdown - this will stop accepting new jobs and allow existing jobs to complete.
		boolean exitted = false;
		try
		{
			// We are willing to wait a little to achieve a clean shutdown.
			exitted = res.awaitTermination(3000, TimeUnit.MILLISECONDS);

		}
		catch (InterruptedException ie)
		{
			Thread.currentThread().interrupt();
		}
		if (!exitted)
		{
			// Some of the running tasks didn't exit in the time we were willing to wait.
			List<Runnable> wereWaiting = res.shutdownNow();  // This will interrupt them if they support interruption.
		}
		LOGGER.log(Level.INFO, "RowcpsExecutorService for {0} shutdown complete.", managerId);
	}
}

File: regi-headless/src/main/python/regi_cli/regi_cli.py
#  Copyright (c) 2026
#  United States Army Corps of Engineers - Hydrologic Engineering Center (USACE/HEC)
#  All Rights Reserved.  USACE PROPRIETARY/CONFIDENTIAL.
#  Source may not be released without written approval from HEC

import os
import logging
import jpype
import jpype.imports
from contextlib import contextmanager
from pathlib import Path

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("regi-launcher")

BASE_DIR = os.path.dirname(os.path.abspath(__file__))
LIB_PATH = os.path.join(BASE_DIR, "lib", "*")
java_home = os.environ.get('JAVA_HOME')
java_bin = os.path.join(java_home, 'bin')
if java_bin not in os.environ['PATH']:
    os.environ['PATH'] = java_bin + os.pathsep + os.environ['PATH']

@contextmanager
def regi_session():
    """
    Context manager to handle JVM lifecycle. 
    Usage:
        with regi_session():
            run_headless(my_func)
    """
    if not jpype.isJVMStarted():
        agent_jar = Path(os.path.join(BASE_DIR, "lib", "opentelemetry-javaagent.jar"))
        agent_flag = f"-javaagent:{agent_jar.absolute()}"

        logger.info(f"Starting JVM with agent: {agent_flag}")
        logger.info("Starting JVM...")
        jpype.startJVM(
            jpype.getDefaultJVMPath(),
            agent_flag,
            convertStrings=True,
            classpath=[LIB_PATH]
        )

    try:
        yield
    finally:
        if jpype.isJVMStarted():
            logger.info("Shutting down JVM...")
            jpype.shutdownJVM()

def run_headless(calculation_callback):
    # We must import these inside the function or after JVM starts
    from usace.rowcps.headless import HeadlessRegiDomainFactory, RegiCalcRegistry
    from usace.rowcps.regi.factories import RowcpsExecutorService
    from java.util.concurrent import TimeUnit
    GlobalOpenTelemetry = jpype.JClass("io.opentelemetry.api.GlobalOpenTelemetry")
    builder = GlobalOpenTelemetry.getTracer("regi-headless").spanBuilder("runHeadless")
    builder.setAttribute("cda.url", os.environ.get("CDA_URL", "unknown"))
    builder.setAttribute("cwms.office", os.environ.get("OFFICE_ID", "unknown"))
    root_span = builder.startSpan()
    scope = root_span.makeCurrent()
    try:
        factory = HeadlessRegiDomainFactory()
        logger.info("Attempting to create RegiDomain...")
        regi_domain = factory.createDomain()

        if regi_domain is not None:
            manager_id = factory.getManagerId()
            registry = RegiCalcRegistry(regi_domain, manager_id)

            try:
                logger.info("Executing callback...")
                calculation_callback(registry)
                regi_domain.commitData(manager_id)
            except Exception as e:
                logger.error("Execution failed.", exc_info=True)
                raise
            finally:
                _shutdown_executor(manager_id)
                regi_domain.closing()
    except (jpype.JException, Exception) as ex:
        # 2. Mirroring the Java catch block
        root_span.recordException(ex)
        StatusCode = jpype.JClass("io.opentelemetry.api.trace.StatusCode")
        root_span.setStatus(StatusCode.ERROR)

        if isinstance(ex, jpype.JException):
            logger.error("Java Exception occurred during headless execution:")
            ex.printStackTrace()
        else:
            logger.error(f"Python Exception occurred during headless execution: {ex}")

        # Mirroring: System.exit(-1)
        scope.close()
        root_span.end()
    finally:
        # 3. Ensuring cleanup
        scope.close()
        root_span.end()

def _shutdown_executor(manager_id):
    from usace.rowcps.regi.factories import RowcpsExecutorService
    from java.util.concurrent import TimeUnit
    res = RowcpsExecutorService.getInstance(manager_id)
    res.shutdown()
    if not res.awaitTermination(3000, TimeUnit.MILLISECONDS):
        res.shutdownNow()